modules/core/src/opengl.cpp

```diff
@@ -819,6 +819,149 @@ unsigned int cv::ogl::Buffer::bufId() const
 #endif
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+// ogl::VertexArray
+#ifndef HAVE_OPENGL
+
+class cv::ogl::VertexArray
+{
+};
+
+#else
+
+class cv::ogl::VertexArray::Impl
+{
+public:
+    static const Ptr<Impl> empty();
+
+    Impl(GLuint vertexArrayId, bool autoRelease);
+    ~Impl();
+
+    void bind() const;
+
+    void setAutoRelease(bool flag) { autoRelease_ = flag; }
+
+    GLuint vertexArrayId() const { return vertexArrayId_; }
+
+    void vertexAttribPointer(GLuint index, GLint size, GLint stride, int offset) const;
+
+private:
+    Impl();
+
+    GLuint vertexArrayId_;
+    bool autoRelease_;
+};
+
+const Ptr<cv::ogl::VertexArray::Impl> cv::ogl::VertexArray::Impl::empty()
+{
+    static Ptr<Impl> p(new Impl);
+    return p;
+}
+
+cv::ogl::VertexArray::Impl::Impl() : vertexArrayId_(0), autoRelease_(false)
+{
+}
+
+cv::ogl::VertexArray::Impl::Impl(GLuint vertexArrayId, bool autoRelease) : vertexArrayId_(vertexArrayId), autoRelease_(autoRelease)
+{
+    gl::GenVertexArrays(1, &vertexArrayId_);
+    CV_CheckGlError();
+
+    CV_Assert(vertexArrayId_ != 0);
+
+    gl::BindVertexArray(vertexArrayId_);
+    CV_CheckGlError();
+
+    gl::BindVertexArray(0);
+    CV_CheckGlError();
+}
+
+cv::ogl::VertexArray::Impl::~Impl()
+{
+    if (autoRelease_ && vertexArrayId_)
+        gl::DeleteVertexArrays(1, &vertexArrayId_);
+}
+
+void cv::ogl::VertexArray::Impl::bind() const
+{
+    gl::BindVertexArray(vertexArrayId_);
+    CV_CheckGlError();
+}
+
+void cv::ogl::VertexArray::Impl::vertexAttribPointer(GLuint index, GLint size, GLint stride, int offset) const
+{
+    gl::BindVertexArray(vertexArrayId_);
+    CV_CheckGlError();
+    gl::VertexAttribPointer(index, size, gl::FLOAT, gl::FALSE_, stride*sizeof(GLfloat), (GLvoid*)(offset * sizeof(GLfloat)));
```

Comment:
I propose to move type to the method parameter with default value and move `*sizeof(GLfloat)` to user side.

Reply:
Like this https://github.com/opencv/opencv/pull/26090/commits/7e78a8557c2a7a441617de696f9ef50661940580 ?